### PR TITLE
feat: add kubernetes example

### PR DESCRIPTION
**Pull request by zw0610**
**What this PR does / why we need it**:

Add your description

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @terrytangyuan 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: test/kubernetes/ftlib-test.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  selector:
+    matchLabels:
+      job: ftlib-test-nrs12
+  template:
+    metadata:
+      labels:
+        job: ftlib-test-nrs12
+    spec:
+      containers:
+        - image: test.caicloudprivatetest.com/library/pytorch-golang:latest
```
Comment (terrytangyuan) [inline test/kubernetes/ftlib-test.yaml]: Switch to use public image

File: test/kubernetes/ftlib-test.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+          - containerPort: 22
+            name: ssh
+          command: ["python", "/crystal/ftlib/ftlib/consensus/gossip/test.py"]
+          resources:
+            limits:
+              nvidia.com/gpu: 1
+          volumeMounts:
+          - name: crystal
+            mountPath: /crystal
+      volumes:
```
Comment (terrytangyuan) [inline test/kubernetes/ftlib-test.yaml]: No need for shared storage?

File: test/kubernetes/ftlib-test.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
+apiVersion: apps/v1
```
Comment (terrytangyuan) [inline test/kubernetes/ftlib-test.yaml]: Are we also aiming to validate that this works with Travis + Minikube?

File: test/kubernetes/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,11 @@
+FROM dockerhub.azk8s.cn/library/golang:1.12.14
```
Comment (terrytangyuan) [inline test/kubernetes/Dockerfile]: Maybe provide an ARG for the location of dockerhub instead of hard-coding this to the mirror in China?

Comment (zw0610) [reply test/kubernetes/Dockerfile]: Right. I shall remove the prefix when this pr is done.

**Comment by terrytangyuan** [comment]
Could we add usages of broadcast and barrier in the example as well?

**Comment by zw0610** [comment]
> Could we add usages of broadcast and barrier in the example as well?

Sure. As a matter of fact, that will be even easier here. I was caught in an issue here for many days where we expect [train_step](https://github.com/caicloud/ftlib/pull/44/files#diff-9bc9d980baa1ec1b98404e43a38e7832R87) function to raise exception when workers lost. However, it is just hanging there.

I will try to fix the issue, otherwise I might just use a broadcast & barrier example without deep learning context here for the moment.

File: test/kubernetes/ftlib-test.yaml (hunk 2 above)
Comment (zw0610) [reply test/kubernetes/ftlib-test.yaml]: removed

File: test/kubernetes/ftlib-test.yaml (hunk 1 above)
Comment (zw0610) [reply test/kubernetes/ftlib-test.yaml]: the image has been replaced.

File: test/kubernetes/ftlib-test.yaml (hunk 3 above)
Comment (zw0610) [reply test/kubernetes/ftlib-test.yaml]: No, I don't think so.

**Comment by zw0610** [comment]
Just for the record, when `nccl` is used as `torch.distributed.backend`, it will hang when worker lost. Therefore, we revert the default backend to `gloo`.

However, I believe it is user's job to guarantee that what was put into `ftlib.execute` will raise exception when worker lost. This part is purely user's logic. (P.S. here the *user* means the user of `ftlib`.)

Also, `gloo` backend is not rather reliable when rebuilding the **process group** in pytorch. Sometimes, it will hang during the `init_process_group`.

**Pull request merged**